### ISSUE TITLE
GDExtension must not make a clone when using reloadable=false

### ISSUE THREAD

**Issue by CycloneRing** [godotengine/godot-proposals#10904]
### Describe the project you are working on

RTX extension for Godot

### Describe the problem or limitation you are having in your project

This is a development wise issue not an actual, Imagine we have a extension "GDExtension.dll" which is NOT a reloadable extension
Now this dll has a dependency "Dependency.dll" which has fallback linking on "GDExtension.dll", This means GDExtension.dll calls to some exported functions of "Dependency.dll" and "Dependency.dll" calls back some exported functions from "GDExtension.dll"

With the [current implementation](https://github.com/godotengine/godot/blob/db66bd35af704fe0d83ba9348b8c50a48e51b2ba/platform/windows/os_windows.cpp#L400) of GDExtension loader it doesn't care if dll is reloadable or not and rename it to ~<Name>.dll ("~GDExtension.dll") regardless of requirement of it. 

And this causes function calls from "Dependency.dll" fail because it's linked against "GDExtension.dll" but "~GDExtension.dll" is loaded in godot.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Issue can be solved very easily, When we parse .gdextension file we have `reloadable = <bool>` so we add a property to 
```CPP
	struct GDExtensionData {
		bool also_set_library_path = false;
		String *r_resolved_path = nullptr;
		bool generate_temp_files = false;
		PackedStringArray *library_dependencies = nullptr;
		bool is_reloadable = false;
	};
```

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Now when we have `is_reloadable` in `GDExtensionData` we simply check if extension is reloadable and "Needs" to be cloned if not we just load the actual extension file.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, It can be fix by less than 10 lines of code.

### Is there a reason why this should be core and not an add-on in the asset library?

Yes, It's the way module loading must be.

**Comment by dsnopek** [comment]
Renaming the DLL on Windows isn't just about supporting reloading. If the DLL isn't renamed, then you are unable to rebuild your extension without stopping the editor first, which makes for an annoying workflow.

For example, some folks don't use reload, but want to keep the editor open while they iterate on their extension, because when they press the play button, the game will launch and load in the latest version of the extension, even if the editor is still using the older one. If the editor has the DLL locked, though, then they can't rebuild it without stopping the editor first.

Perhaps could add an editor setting to disable the naming?

Also, just to make sure I understand your use case: you've got another extension which is linked against the first, and the renaming means that the dynamic linker is loading the wrong DLL?

**Comment by CycloneRing** [comment]
@dsnopek Sorry but it doesn't make sense to me, Why you need to rebuild the project when you're not using hot-reloading? Do you mean like build and hit play so debug mode will read original dll directly?

However, the renaming is the issue here, Instead of adding ~ to dll name it can be copied to cache directory or a folder as temp beside actual dll, The issue is name of a dll, It must not be changed, Dll names are very important, imagine if d3d11.dll be renamed to d3d11_.dll, Now same happens in my case, I don't have another extension it's just another library that calls back to extension.

Since godot api are only solved inside a GDExtension dll a dependecny dll cannot directly call or use godot-cpp api, Instead it should use wrapper functions provided by loaded GDExtension dll.

For example, We have a gdextension dll named Terrain.dll and this Terrain.dll imports a function from a dependency TerrainMath.dll now if this import be one-way from Terrain.dll to TerrainMath.dll everything works the way it is. Now what if we have a UpdateTerrainData() function exported from Terrain.dll which updates computed terrain data we can't link against it in TerrainMath.dll. Because TerrainMath.dll  is linked back to Terrain.dll and godot loaded ~Terrain.dll 
So two things happens :
- TerrainMath.dll looks for Terrain.dll and it can't find it, It will crash.
-  TerrainMath.dll looks for Terrain.dll and it finds it, It loads it but It will crash because it's not loaded by GDExtension loader and functions are not solved.

Only thing important to fix the issue is the dll name untouched, It can be loaded from a alternative place.

**Comment by dsnopek** [comment]
> Why you need to rebuild the project when you're not using hot-reloading? Do you mean like build and hit play so debug mode will read original dll directly?

Yes. Keeping the editor open, rebuilding the DLL, and hitting play so that the game will load the new DLL (even though the old DLL is still loaded in the editor because reload is disabled).
